exporter/awsemfexporter/cwlog_client.go

```diff
@@ -0,0 +1,244 @@
+//if necessary.
+func (client *CloudWatchLogClient) PutLogEvents(input *cloudwatchlogs.PutLogEventsInput, retryCnt int) *string {
+	var response *cloudwatchlogs.PutLogEventsOutput
+	var err error
+	var token = input.SequenceToken
+
+	for i := 0; i <= retryCnt; i++ {
+		input.SequenceToken = token
+		response, err = client.svc.PutLogEvents(input)
+		if err != nil {
+			awsErr, ok := err.(awserr.Error)
+			if !ok {
+				client.logger.Error(fmt.Sprintf("E! Cannot cast PutLogEvents error %#v into awserr.Error.", err))
+				return token
+			}
+			switch e := awsErr.(type) {
+			case *cloudwatchlogs.InvalidParameterException:
+				client.logger.Error(fmt.Sprintf("E! cloudwatchlogs: %s for log group %s log stream %s, will not retry the request: %s, original error: %#v, %#v",
+					e.Code(),
+					*input.LogGroupName,
+					*input.LogStreamName,
+					e.Message(),
+					e.Error(),
+					e))
+				return token
+			case *cloudwatchlogs.InvalidSequenceTokenException: //Resend log events with new sequence token when InvalidSequenceTokenException happens
+				client.logger.Warn(fmt.Sprintf("W! cloudwatchlogs: %s, will search the next token and retry the request: %s, original error: %#v, %#v",
+					e.Code(),
+					e.Message(),
+					e.Error(),
+					e))
+				backoffSleep(i)
+				token = e.ExpectedSequenceToken
+				continue
+			case *cloudwatchlogs.DataAlreadyAcceptedException: //Skip batch if DataAlreadyAcceptedException happens
+				client.logger.Warn(fmt.Sprintf("W! cloudwatchlogs: %s, drop this request and continue to the next request: %s, original error: %#v, %#v",
+					e.Code(),
+					e.Message(),
+					e.Error(),
+					e))
+				token = e.ExpectedSequenceToken
+				return token
+			case *cloudwatchlogs.OperationAbortedException: //Retry request if OperationAbortedException happens
+				client.logger.Warn(fmt.Sprintf("W! cloudwatchlogs: %s, will retry the request: %s, original error: %#v, %#v",
+					e.Code(),
+					e.Message(),
+					e.Error(),
+					e))
+				backoffSleep(i)
+				continue
+			case *cloudwatchlogs.ServiceUnavailableException: //Retry request if ServiceUnavailableException happens
+				client.logger.Warn(fmt.Sprintf("W! cloudwatchlogs: %s, will retry the request: %s, original error: %#v, %#v",
+					e.Code(),
+					e.Message(),
+					e.Error(),
+					e))
+				backoffSleep(i)
+				continue
+			case *cloudwatchlogs.ResourceNotFoundException:
+				tmpToken, tmpErr := client.CreateStream(input.LogGroupName, input.LogStreamName)
+				if tmpErr == nil {
+					if tmpToken == "" {
+						token = nil
+					} else {
+						token = &tmpToken
+					}
+				}
+				backoffSleep(i)
+				continue
+			default:
+				// ThrottlingException is handled here because the type cloudwatch.ThrottlingException is not yet available in public SDK
+				// Retry request if ThrottlingException happens
+				if awsErr.Code() == ErrCodeThrottlingException {
```

Comment:
Retrying during throttling will never catch up. Just drop the data at this point. Possibly even the entire sequence.